Fix: Supports Unicode BOM (fixes #4878)

mysticatea opened the pull request:
Fixes #4878.

- `eslint.verify()` API came to strip BOM before parsing.
- `hasBOM` property was added into `SourceCode` object.
- `SourceCodeFixer.applyFixes()` came to insert BOM to the head of output if `sourceCode.hasBOM` is `true`.

> This diff becomes more readable if you add `?w=1`.

mysticatea commented on lib/eslint.js:
```diff
@@ -818,6 +825,13 @@ module.exports = (function() {
+        // If there is BOM then it increases the column of the first line.
+        if (sourceCode && sourceCode.hasBOM) {
+            for (var i = 0; i < messages.length && messages[i].line === 1; ++i) {
+                messages[i].column += 1;
```
I'm not sure this is needed.

nzakas replied:
I don't think so. The BOM is zero-width, so people won't see it in editors.

mysticatea commented on tests/lib/eslint.js:
```diff
@@ -2112,6 +2112,113 @@ describe("eslint", function() {
+        });
+
+        // Skip in browsers because we cannot use "fs".
+        (typeof window === "object" ? xdescribe : describe)("when it read a UTF-8 file (has BOM), SourceCode", function() {
```
Should I separate test file for this?

The purpose of this test is to confirm that it can handle UTF-8 BOM `[0xEF, 0xBB, 0xBF]` in a file.
In other tests, those are using UTF-16 BOM in a string literal.

nzakas commented on lib/eslint.js:
```diff
@@ -673,10 +673,17 @@ module.exports = (function() {
             // there's no input, just exit here
             if (text.trim().length === 0) {
-                sourceCode = new SourceCode(text, blankScriptAST);
+                sourceCode = new SourceCode(text, blankScriptAST, {hasBOM: hasBOM});
```
I think we need to keep the text as it was originally when we pass it in here sonthe BOM is present. Then, you can calculate `hasBOM` in the constructor.

mysticatea commented:
I updated this PR.

- Removed the third argument of `SourceCode` constructor.
- Added the code which calculates `hasBOM` into `SourceCode` constructor.
- Moved several tests from `tests/lib/eslint.js` to `tests/lib/util/source-code.js`.

`eslint.verify()` still is stripping BOM before parsing.

nzakas commented on lib/eslint.js:
```diff
@@ -673,17 +686,22 @@ module.exports = (function() {
             if (text.trim().length === 0) {
-                sourceCode = new SourceCode(text, blankScriptAST);
+                sourceCode = new SourceCode(
+                    stripUnicodeBOM(text),
```
This should be `text`, so `SourceCode` has the original text.

mysticatea replied:
Oh this is a mistake!

nzakas commented on lib/util/source-code.js:
```diff
@@ -87,19 +87,28 @@ function looksLikeExport(astNode) {
      * @type string
      */
-    this.text = text;
+    this.text = (this.hasBOM ? text.slice(1) : text);
```
Do we need to strip the BOM at all here? What if we just left this as `text`?

mysticatea replied:
But AST is created from the stripped text.
BOM is zero width character but 1 character.
So I guessed AST gets wrong `range` if this is original text.

nzakas replied:
Ah, I see. Thanks for explaining.

mysticatea commented:
I updated this PR.

- Fixed https://github.com/eslint/eslint/pull/4935#discussion_r49769347

nzakas commented:
Can you update the `SourceCode` docs? https://github.com/eslint/eslint/blob/master/docs/developer-guide/nodejs-api.md#sourcecode

mysticatea commented:
Sure, I had been forgetting docs.

mysticatea commented:
I updated docs.

If a given AST object was parsed from not stripped text, the source code object might get wrong range of AST. So I changed this commit tag to `Breaking`.

And I updated source-code-fixer.
Now `fixer.removeRange([-1, 0])` removes BOM, `fixer.insertBeforeRange([0, 0], "\uFEFF")` inserts BOM.

nzakas commented:
Lgtm.

platinumazure commented on tests/lib/util/source-code-fixer.js:
```diff
@@ -242,6 +270,221 @@ describe("SourceCodeFixer", function() {
+                assert.equal(result.messages.length, 0);
+            });
+
+            it("should insert BOM with an insertion of '\uFEFFfoobar' at 0", function() {
```
Seems the test description doesn't match what actually happens here... (foobar vs // start)

mysticatea replied:
My intention here, foobar is "any string".

platinumazure replied:
Fair enough, but having foobar in the quotes makes it look like a string
literal so I assumed that was what you were aiming for. Maybe it could say
something like "... '\uxFEFF' plus any string..." so that only the BOM is
in the string literal? Then again maybe I'm overthinking it.
On Jan 18, 2016 12:51 PM, "Toru Nagashima" <notifications@github.com> wrote:

> In tests/lib/util/source-code-fixer.js
> <https://github.com/eslint/eslint/pull/4935#discussion_r50032422>:
>
> > @@ -242,6 +270,221 @@ describe("SourceCodeFixer", function() {
> >
> >          });
> >
> > +        describe("BOM manipulations", function() {
> > +
> > +            it("should insert BOM with an insertion of '\uFEFF' at 0", function() {
> > +                var result = SourceCodeFixer.applyFixes(sourceCode, [ INSERT_BOM ]);
> > +                assert.equal(result.output, "\uFEFF" + TEST_CODE);
> > +                assert.isTrue(result.fixed);
> > +                assert.equal(result.messages.length, 0);
> > +            });
> > +
> > +            it("should insert BOM with an insertion of '\uFEFFfoobar' at 0", function() {
>
> My intention here, foobar is "any string".
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/eslint/eslint/pull/4935/files#r50032422>.
>

mysticatea replied:
Thank you for the advice. I learnt it. I had not gotten that phrase before.

platinumazure commented on tests/lib/util/source-code-fixer.js:
```diff
@@ -242,6 +270,221 @@ describe("SourceCodeFixer", function() {
+                assert.equal(result.messages.length, 0);
+            });
+
+            it("should replace BOM with a negative range and 'foobar'", function() {
```
Seems the test description doesn't match what actually happens here... (foobar vs // start)
